Fix regression in ConnectAttempt

Bronek opened the pull request:
<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.

If your branch is on a personal fork and has a name that allows it to
run CI build/test jobs (e.g. "ci/foo"), remember to rename it BEFORE
opening the PR.  This avoids unnecessary redundant test runs. Renaming
the branch after opening the PR will close the PR.
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/renaming-a-branch
-->

## High Level Overview of Change

Fix a regression bug from #5669 

<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

A regression was introduced in #5669 which would cause `rippled` to potentially dereference a disengaged `std::optional` when connecting to a peer. This would cause UB in release build and crash in debug.

The bug does not affect version 3.0.0 since #5669 was not planned for release in this version and got explicitly reverted.

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a spec or design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [x] Bug fix (non-breaking change which fixes an issue)

Tapanito reviewed (approved):
Great catch! Thanks Bronek.

vvysokikh1 commented:
@Tapanito the semantics of optional and string seem to be really messed up here.
I'm a bit concerned that the change here fixes the crash but changes the meaning of the check (checking if string is empty vs checking if optional has value). Should both these conditions be checked here? Should it be purely std::string (without optional)?

Tapanito commented:
@vvysokikh1 the previous code simply did a cast to bool, ` auto const result = overlay_.peerFinder().activate(slot_, publicKey, static_cast<bool>(member));`, which I understand is checking if the member (`std::optional`) has a value.

Bronek commented:
> @vvysokikh1 the previous code simply did a cast to bool, ` auto const result = overlay_.peerFinder().activate(slot_, publicKey, static_cast<bool>(member));`, which I understand is checking if the member (`std::optional`) has a value.

The regression that's being fixed here is that #5669 **did** change the semantics, from checking an equivalent of `member.has_value()` (indirectly, via [cast to `bool`](https://en.cppreference.com/w/cpp/utility/optional/operator_bool.html)) to `empty()` on a string, unconditionally dereferenced. Here we are bringing the old semantics back.

Here's the old code : https://github.com/XRPLF/rippled/blob/ccb9f1e42d13bf83c9c953fb562938e3734a699e/src/xrpld/overlay/detail/ConnectAttempt.cpp#L409

vvysokikh1 commented:
> @vvysokikh1 the previous code simply did a cast to bool, ` auto const result = overlay_.peerFinder().activate(slot_, publicKey, static_cast<bool>(member));`, which I understand is checking if the member (`std::optional`) has a value.

yes, and you changed that to check if the string is not empty. My question here - was that an intentional change or a mistake. If that was intentional, this fix is not sufficient. If by mistake - i can approve then

Bronek commented:
> > @vvysokikh1 the previous code simply did a cast to bool, ` auto const result = overlay_.peerFinder().activate(slot_, publicKey, static_cast<bool>(member));`, which I understand is checking if the member (`std::optional`) has a value.
> 
> yes, and you changed that to check if the string is not empty. My question here - was that an intentional change or a mistake. If that was intentional, this fix is not sufficient. If by mistake - i can approve then

FWIW I find nothing in #5669 to indicate that was intentional. There's no change in `PeerfinderManager.cpp` or in `Cluster.cpp` that would indicate any change of semantics to the return value of `app_.cluster().member(publicKey);` or its use in `overlay_.peerFinder().activate(...)` or explanation how that might be possibly related to SSL socket shutdown.

bthomee committed:
fix: Fix regression in ConnectAttempt

A regression was introduced in #5669 which would cause rippled to potentially dereference a disengaged std::optional when connecting to a peer. This would cause UB in release build and crash in debug.

Bronek commented:
@Tapanito the embarrassing part is that, when merging #5669 I did look at that part and did not read it carefully enough 🥹
